lib/ClangToAst.ml

```diff
@@ -1396,6 +1482,63 @@ let rec translate_stmt (env : env) (s : Clang.Ast.stmt_desc) : Krml.Ast.expr =
     end
   | If { cond = { desc = BinaryOperator { lhs; kind = NE; rhs }; _ }; then_branch; _ }
     when has_pointer_type lhs && is_null rhs -> translate_stmt env then_branch.desc
+
+  | If { cond; then_branch; else_branch; _ } when is_tag_check env cond ->
+      let var, variant, case_ref = deconstruct_tag_check env cond in
```

Comment:
this is the crux of the PR: it would help to have a pseudocode comment that explains that `if (x.tag == foo) { ... } else { ... }` gets rewritten to `match x with | Foo ...` (or whatever the rewriting is)